Peer Node Crash When Executing ChainCode

Fabric version: 2.4.6

When we invoked chaincode from our program the related peer node crash with the following error: 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1580fc9]

goroutine 114930 [running]:
github.com/hyperledger/fabric/internal/pkg/gateway.rwsetDifference({0xc002d9fe40?, 0x142?, 0x160?}, {0xc002f18000, 0x142, 0x160})
	/go/src/github.com/hyperledger/fabric/internal/pkg/gateway/apiutils.go:325 +0x1c89
github.com/hyperledger/fabric/internal/pkg/gateway.payloadDifference({0xc000025860?, 0xc002687d30?, 0xc002687d10?}, {0xc000025a40, 0x1c3, 0x1e0})
	/go/src/github.com/hyperledger/fabric/internal/pkg/gateway/apiutils.go:278 +0x145
github.com/hyperledger/fabric/internal/pkg/gateway.(*plan).processEndorsement(0xc002e6f380, 0xc002eb43c0, 0xc003085f20)
	/go/src/github.com/hyperledger/fabric/internal/pkg/gateway/endorsement.go:123 +0x5fe
github.com/hyperledger/fabric/internal/pkg/gateway.(*Server).processProposal(0xc00268c300, {0x1b56198?, 0xc00288eb70}, 0xc002e6f100?, 0xc002eb43c0, 0xc002e600f0, 0xc0035f2038)
	/go/src/github.com/hyperledger/fabric/internal/pkg/gateway/api.go:263 +0x62c
github.com/hyperledger/fabric/internal/pkg/gateway.(*Server).Endorse.func1(0xc00029f750?)
	/go/src/github.com/hyperledger/fabric/internal/pkg/gateway/api.go:190 +0xa6
created by github.com/hyperledger/fabric/internal/pkg/gateway.(*Server).Endorse
	/go/src/github.com/hyperledger/fabric/internal/pkg/gateway/api.go:188 +0xd75
```

Any idea which area we can check to fix the error?

Thanks in advanced any guide and help given.  🙏🏻

Cheers.

--- a/internal/pkg/gateway/apiutils.go
+++ b/internal/pkg/gateway/apiutils.go
@@ -312,7 +312,7 @@ func rwsetDifference(rwset1, rwset2 []byte) (*resultDifference, error) {
 	summarySet := nsRWsets{}
 	rwDiff := &resultDifference{}
 
-	for _, txrw := range txrw1.NsRwset {
+	for _, txrw := range txrw1.GetNsRwset() {
 		reads := readset{}
 		writes := writeset{}
 		pvtHashes := writeset{}
@@ -321,72 +321,72 @@ func rwsetDifference(rwset1, rwset2 []byte) (*resultDifference, error) {
 		if err != nil {
 			return nil, err
 		}
-		for _, r := range kvrws.Reads {
-			reads[r.Key] = r.Version.BlockNum
+		for _, r := range kvrws.GetReads() {
+			reads[r.GetKey()] = r.GetVersion().GetBlockNum()
 		}
-		for _, w := range kvrws.Writes {
-			writes[w.Key] = w.Value
+		for _, w := range kvrws.GetWrites() {
+			writes[w.GetKey()] = w.GetValue()
 		}
-		for _, mw := range kvrws.MetadataWrites {
+		for _, mw := range kvrws.GetMetadataWrites() {
 			entryset := writeset{}
-			for _, me := range mw.Entries {
-				entryset[me.Name] = me.Value
+			for _, me := range mw.GetEntries() {
+				entryset[me.GetName()] = me.GetValue()
 			}
-			metadata[mw.Key] = entryset
+			metadata[mw.GetKey()] = entryset
 		}
 		for _, chrws := range txrw.GetCollectionHashedRwset() {
-			pvtHashes[chrws.CollectionName] = chrws.PvtRwsetHash
+			pvtHashes[chrws.GetCollectionName()] = chrws.GetPvtRwsetHash()
 		}
-		summarySet[txrw.Namespace] = readwriteset{r: reads, w: writes, m: metadata, p: pvtHashes}
+		summarySet[txrw.GetNamespace()] = readwriteset{r: reads, w: writes, m: metadata, p: pvtHashes}
 	}
-	for _, txrw := range txrw2.NsRwset {
+	for _, txrw := range txrw2.GetNsRwset() {
 		var reads readset
 		var writes writeset
 		var pvtHashes writeset
 		var metadata metaset
-		if rw, ok := summarySet[txrw.Namespace]; ok {
+		if rw, ok := summarySet[txrw.GetNamespace()]; ok {
 			reads = rw.r
 			writes = rw.w
 			metadata = rw.m
 			pvtHashes = rw.p
 		}
-		kvrws, err := protoutil.UnmarshalKVRWSet(txrw.Rwset)
+		kvrws, err := protoutil.UnmarshalKVRWSet(txrw.GetRwset())
 		if err != nil {
 			return nil, err
 		}
-		for _, r := range kvrws.Reads {
-			block := reads[r.Key] // missing entry will be represented by the zero value
-			if block != r.Version.BlockNum {
+		for _, r := range kvrws.GetReads() {
+			block := reads[r.GetKey()] // missing entry will be represented by the zero value
+			if block != r.GetVersion().GetBlockNum() {
 				// state is at different version (or not present in rwset1 if block is zero)
-				rwDiff.addReadDiff(txrw.Namespace, r.Key, block, r.Version.BlockNum)
+				rwDiff.addReadDiff(txrw.GetNamespace(), r.GetKey(), block, r.GetVersion().GetBlockNum())
 			}
-			delete(reads, r.Key)
+			delete(reads, r.GetKey())
 		}
-		for _, w := range kvrws.Writes {
-			value := writes[w.Key]
-			if !bytes.Equal(value, w.Value) {
+		for _, w := range kvrws.GetWrites() {
+			value := writes[w.GetKey()]
+			if !bytes.Equal(value, w.GetValue()) {
 				// state writes different value (or not present in rwset1 if value is nil)
-				rwDiff.addWriteDiff(txrw.Namespace, w.Key, value, w.Value)
+				rwDiff.addWriteDiff(txrw.GetNamespace(), w.GetKey(), value, w.GetValue())
 			}
-			delete(writes, w.Key)
+			delete(writes, w.GetKey())
 		}
-		for _, mw := range kvrws.MetadataWrites {
-			expected := metadata[mw.Key]
-			for _, e := range mw.Entries {
-				value := expected[e.Name]
-				if !bytes.Equal(value, e.Value) {
-					rwDiff.addMetadataWriteDiff(txrw.Namespace, mw.Key, e.Name, value, e.Value)
+		for _, mw := range kvrws.GetMetadataWrites() {
+			expected := metadata[mw.GetKey()]
+			for _, e := range mw.GetEntries() {
+				value := expected[e.GetName()]
+				if !bytes.Equal(value, e.GetValue()) {
+					rwDiff.addMetadataWriteDiff(txrw.GetNamespace(), mw.GetKey(), e.GetName(), value, e.GetValue())
 				}
-				delete(expected, e.Name)
+				delete(expected, e.GetName())
 			}
 		}
 		for _, chrws := range txrw.GetCollectionHashedRwset() {
-			hash := pvtHashes[chrws.CollectionName]
-			if !bytes.Equal(hash, chrws.PvtRwsetHash) {
+			hash := pvtHashes[chrws.GetCollectionName()]
+			if !bytes.Equal(hash, chrws.GetPvtRwsetHash()) {
 				// state writes different value (or not present in rwset1 if value is nil)
-				rwDiff.addPvtHashDiff(txrw.Namespace, chrws.CollectionName, hash, chrws.PvtRwsetHash)
+				rwDiff.addPvtHashDiff(txrw.GetNamespace(), chrws.GetCollectionName(), hash, chrws.GetPvtRwsetHash())
 			}
-			delete(pvtHashes, chrws.CollectionName)
+			delete(pvtHashes, chrws.GetCollectionName())
 		}
 	}
 	// whatever is left in the summary set is present in rwset1 but not rwset2

--- a/internal/pkg/gateway/apiutils_test.go
+++ b/internal/pkg/gateway/apiutils_test.go
@@ -7,6 +7,7 @@ SPDX-License-Identifier: Apache-2.0
 package gateway
 
 import (
+	"math"
 	"testing"
 
 	"github.com/hyperledger/fabric-protos-go/common"
@@ -168,6 +169,43 @@ func TestPayloadDifferenceReadExtra(t *testing.T) {
 	expected := [][]interface{}{
 		{"type", "extraneous read", "namespace", "ns1", "key", "key3", "initial-endorser-value", "0", "invoked-endorser-value", "3"},
 		{"type", "extraneous read", "namespace", "ns2", "key", "key3b", "initial-endorser-value", "0", "invoked-endorser-value", "5"},
+	}
+	require.ElementsMatch(t, expected, diff.details())
+}
+
+func TestPayloadDifferenceReadMissingProtos(t *testing.T) {
+	rpl1 := createProposalResponsePayload(
+		t, &responseT{payload: []byte("my_value"), status: 200, message: "no error"},
+		[]*readT{
+			{namespace: "ns1", key: "key1", block: math.MaxInt64},
+			{namespace: "ns1"},
+		},
+		[]*writeT{},
+		[]*metaWriteT{},
+		[]*pvtCollectionT{},
+		nil,
+	)
+
+	rpl2 := createProposalResponsePayload(
+		t, &responseT{payload: []byte("my_value"), status: 200, message: "no error"},
+		[]*readT{
+			{namespace: "ns1", key: "key1", block: 4},
+		},
+		[]*writeT{},
+		[]*metaWriteT{},
+		[]*pvtCollectionT{},
+		nil,
+	)
+
+	rpl1Bytes := marshal(rpl1, t)
+	rpl2Bytes := marshal(rpl2, t)
+
+	diff, err := payloadDifference(rpl1Bytes, rpl2Bytes)
+	require.NoError(t, err)
+
+	expected := [][]interface{}{
+		{"type", "extraneous read", "namespace", "ns1", "key", "key1", "initial-endorser-value", "0", "invoked-endorser-value", "4"},
+		{"type", "extraneous read", "namespace", "ns1", "key", "", "initial-endorser-value", "0", "invoked-endorser-value", "0"},
 	}
 	require.ElementsMatch(t, expected, diff.details())
 }
@@ -468,9 +506,17 @@ func collateReadWriteSets(t *testing.T, reads []*readT, writes []*writeT, metaWr
 			rwset = &kvrwset.KVRWSet{}
 			grouped[r.namespace] = rwset
 		}
+		if r.key == "" && r.block == 0 { // signifies nil Read proto in these tests
+			rwset.Reads = append(rwset.Reads, nil)
+			continue
+		}
+		var version *kvrwset.Version
+		if r.block != math.MaxInt64 { // signifies missing version in these tests
+			version = &kvrwset.Version{BlockNum: r.block}
+		}
 		rwset.Reads = append(rwset.Reads, &kvrwset.KVRead{
 			Key:     r.key,
-			Version: &kvrwset.Version{BlockNum: r.block},
+			Version: version,
 		})
 	}
 	for _, w := range writes {

--- a/internal/pkg/gateway/event/chaincode.go
+++ b/internal/pkg/gateway/event/chaincode.go
@@ -19,15 +19,15 @@ func (event *ChaincodeEvent) Transaction() *Transaction {
 }
 
 func (event *ChaincodeEvent) ChaincodeID() string {
-	return event.message.ChaincodeId
+	return event.message.GetChaincodeId()
 }
 
 func (event *ChaincodeEvent) EventName() string {
-	return event.message.EventName
+	return event.message.GetEventName()
 }
 
 func (event *ChaincodeEvent) Payload() []byte {
-	return event.message.Payload
+	return event.message.GetPayload()
 }
 
 func (event *ChaincodeEvent) ProtoMessage() *peer.ChaincodeEvent {

